### ISSUE TITLE
CephObjectStoreUser Reference Secret

### ISSUE THREAD

**Issue by jameshearttech** [rook/rook#11563]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
* CephObjectStoreUser should optionally be able to reference a secret where S3 key is defined. This enables us to specify the accesskey and accesssecret rather than those values being randomly generated. When the secret changes the CephObjectStoreUser should be updated.

**What is use case behind this feature:**
* Define CephObjectStoreUser S3 key properties (e.g., accesskey and accesssecret) so we can use those same values to configure access to Ceph object storage bucket in an automated way (e.g., using ArgoCD and External Secrets Operator).

**Environment**:
```
$ kubectl get nodes -o wide
NAME          STATUS   ROLES           AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
dev-master0   Ready    control-plane   2d23h   v1.26.0   10.69.2.30    <none>        Ubuntu 22.04.1 LTS   5.15.0-58-generic   containerd://1.6.15
dev-master1   Ready    control-plane   2d23h   v1.26.0   10.69.2.31    <none>        Ubuntu 22.04.1 LTS   5.15.0-58-generic   containerd://1.6.15
dev-master2   Ready    control-plane   2d23h   v1.26.0   10.69.2.32    <none>        Ubuntu 22.04.1 LTS   5.15.0-58-generic   containerd://1.6.15
dev-worker0   Ready    <none>          2d23h   v1.26.0   10.69.2.33    <none>        Ubuntu 22.04.1 LTS   5.15.0-58-generic   containerd://1.6.15
dev-worker1   Ready    <none>          2d23h   v1.26.0   10.69.2.34    <none>        Ubuntu 22.04.1 LTS   5.15.0-58-generic   containerd://1.6.15
dev-worker2   Ready    <none>          2d23h   v1.26.0   10.69.2.35    <none>        Ubuntu 22.04.1 LTS   5.15.0-58-generic   containerd://1.6.15
dev-worker3   Ready    <none>          2d23h   v1.26.0   10.69.2.36    <none>        Ubuntu 22.04.1 LTS   5.15.0-58-generic   containerd://1.6.15
```

**Comment by divaspathak** [comment]
Hello @jameshearttech, I want to work on this issue. Can you please assign it to me?

**Comment by subhamkrai** [comment]
@jameshearttech we do store `accesskey` and `accesssecret`. what are you requesting here? you suggesting to objectstoreuser should point to someSecret for this values?

**Comment by parth-gr** [comment]
Probably they are asking similar to this https://github.com/rook/rook/pull/11336

**Comment by thotz** [comment]
> Probably they are asking similar to this #11336

Yes I guess so, @jameshearttech can you please check above PR owner about its status?

**Comment by jameshearttech** [comment]
[https://rook-io.slack.com/archives/CK9CF5H2R/p1674075543129569](https://rook-io.slack.com/archives/CK9CF5H2R/p1674075543129569)

In my case everything is committed to Git. Storing the accesskey and accesssecret in the CephObjectUserStore resource would potentially leak those sensitive values.

> @jameshearttech we do store `accesskey` and `accesssecret`. what are you requesting here? you suggesting to objectstoreuser should point to someSecret for this values?

Yes, exactly.

**Comment by zhucan** [comment]
is there any new progress? @jameshearttech  @thotz  @divaspathak @subhamkrai @parth-gr

**Comment by thotz** [comment]
I thought this issue was related to https://github.com/rook/rook/pull/11336

**Comment by jameshearttech** [comment]
Yeah, they are definitely related. If it makes sense, feel free to close this one.

**Comment by devops-42** [comment]
Hi @jameshearttech,

I'd appreciate the approach of referring to existing k8s secrets. This would decouple the direct communication from Ceph to a KMS like Vault. One could implement K8s secrets by themselve and create `CephObjectStoreUser` instances referencing to an existing K8s secret.

**Comment by jhoblitt** [comment]
I'm interested in this feature as well. I have rgw clients which are external to k8s. It would be great if I could using something like external-secrets operator to sync rgw user creds. This would be an alternative to using ldap with rgw.

**Comment by jhoblitt** [comment]
This is a feature request we should probably keep open.

**Comment by jameshearttech** [comment]
Please do not close feature request.

**Comment by travisn** [comment]
A few notes on implementing this feature for someone to pick it up:
In the [createOrUpdateCephUser()](https://github.com/rook/rook/blob/4b701da916c74282898d8b8fc3dfa32765bd1f2d/pkg/operator/ceph/object/user/controller.go#L290) method, on line 298 if the `GetUser()` returns `ErrNoSuchUser`, then we need to:
1. Query if the K8s secret exists where Rook would later store the user creds. 
2. If the secret exists and the expected keys exist (see example below), then set those values on the `r.userConfig` struct that is being passed to the `CreateUser()` method. 
3. Assuming Ceph accepts the user secret as input, the user would be created with the pre-defined keys.

Here is an example user secret which is named `rook-ceph-object-user-<store>-<user>`:
```yaml
% kubectl -n rook-ceph get secret rook-ceph-object-user-my-store-my-user -o yaml
apiVersion: v1
kind: Secret
data:
  AccessKey: U0lVNVJQTVg0VFIwMTI1QkxGSjc=
  SecretKey: aHdnbGdIVmlDcEJqTHN3WUJDakJqWlBRQ2tzeEM3dkRuRDdXMTNySw==
  Endpoint: aHR0cDovL3Jvb2stY2VwaC1yZ3ctbXktc3RvcmUucm9vay1jZXBoLnN2Yzo4MA==
```

If this secret is created in advance of the reconcile, then Rook would pick it up.

Note that since this issue was originally opened, now the user and secret could be created in any namespace since #12730 was implemented.

**Comment by subhamkrai** [comment]
Removing the assigning as I'm focused on some other issue and I don't want to hold this to my name for long

**Comment by subhamkrai** [comment]
Also added it to project board 1.13 to get more attention

**Comment by BlaineEXE** [comment]
I'd like to discuss/plan some implementation details here before continuing with the PR.

Firstly, I think this work is worthwhile. It has been something that I've wanted for quite some time to help make our RGW multisite CI tests simpler. However, there are complexities that we should discuss and have a detailed plan for. Otherwise, the implementation PR will need to be rewritten each time we discuss a new detail.

#### First detail: Source of truth

Rook generally defers to Ceph to be the source of truth for many things, including auth keys. That means that today, the way to rotate keys would be to use `radosgw-admin` or the RGW "Admin Ops API". After that, restarting the Rook operator would update the k8s secret.

I can see how that workflow wouldn't work or be desired for all users. However, on the flip side, I can see how many users might want to use this workflow. For example, maybe they use the Ceph dashboard to rotate the creds.

My suggestion for how to resolve this is to use the current behavior as the default behavior. For users who want to treat the secret itself as the source of truth, my thought is to use an annotation on the secret to inform Rook's internal behavior. Perhaps an annotation on the secret like this would instruct Rook to treat the secret as the source of truth: `administration.rook.io/source-of-truth: secret`.

While we could use a boolean on the CRD to select this behavior, it seems more natural to me to use an annotation on the secret itself since the administrator can easily apply the annotation at the same time they create/update the secret with their desired key. 

#### Second detail: Key array

Parth has pointed out that RGW internally has an array of keys for the object store. I think this helps to provide a time window for applications to migrate from the old key to the new key. This behavior could be important and desired for some administrators.

However, some administrators might want to revoke the old key and add the new key immediately with no overlap time. This poses a bit of a headache for the design.

The current secret looks like this:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-object-user-my-store-my-user
  namespace: rook-ceph
data:
  AccessKey: ACCESS
  SecretKey: SECRET
```

In order to keep backwards compatibility, the secret's fields cannot change. My proposal would be to add a field that is a json-encoded list of all secret keys. When a user has multiple keys, Rook will take the last key in the list as `SecretKey` (with the assumption that the last key in the list is most likely to be the newer and longer-lived key), as shown below:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-object-user-my-store-my-user
  namespace: rook-ceph
data:
  AccessKey: ACCESS
  SecretKey: SECRET2
  SecretKeys: `["SECRET1", "SECRET2"]`
```

This example assumes that the admin has added a secret key to the user after Rook initially created the user.

#### Combined details

With both of these details in mind, my proposal is as such:

For administrators who want to control the keys via the secret itself, they would apply the annotation and update the keys as desired.

Admins can change `SecretKey` and add it into the `SecretKeys` array as shown.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-object-user-my-store-my-user
  namespace: rook-ceph
  annotations:
    - administration.rook.io/source-of-truth: secret
data:
  AccessKey: ACCESS
  SecretKey: ADMINSECRET
  SecretKeys: `["SECRET", "ADMINSECRET"]`
```

When they are ready, they can remove the old secret from the `SecretKeys` array by updating it as shown.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-object-user-my-store-my-user
  namespace: rook-ceph
  annotations:
    - administration.rook.io/source-of-truth: secret
data:
  AccessKey: ACCESS
  SecretKey: ADMINSECRET
  SecretKeys: `["ADMINSECRET"]`
```

An admin could decide to change `SecretKey` and remove the old key in the `SecretKeys` array at the same time if they don't want a transition period.

When the source-of-truth annotation is set, Rook should ensure that the `SecretKey` is present in the `SecretKeys` array. If it isn't, Rook must error on object store reconciliation with a message alerting the user to the discrepancy. Rook has no way of knowing which value has the correct field, so it is up to the user to fix it.

**Comment by BlaineEXE** [comment]
> If the annotation is set, Rook is the source of truth, so shouldn't Rook update the entire array of keys to match what is in the K8s secret?

Yes. That is what I intended. But there is a chance of user error with both `SecretKey` and `SecretKeys` present. Rook should make sure that `SecretKey` is in the list of `SecretKeys`. If it isn't, then the user has made an error, and Rook can't know for sure whether `SecretKey` is right or `SecretKeys` is right.

**Comment by travisn** [comment]
> > If the annotation is set, Rook is the source of truth, so shouldn't Rook update the entire array of keys to match what is in the K8s secret?
> 
> Yes. That is what I meant to say. But there is a chance of user error with both `SecretKey` and `SecretKeys` present. Rook should make sure that `SecretKey` is in the list of `SecretKeys`. If it isn't, then the user has made an error, and Rook can't know for sure whether `SecretKey` is right or `SecretKeys` is right.

What if we take the union of `SecretKey` and `SecretKeys`? Seems like the union is the desired state in case they forget to include the main SecretKey in the list of SecretKeys.

**Comment by jhoblitt** [comment]
I am not enthusiastic about encoding json into strings (enough though this is a common pattern with annotations) as would mean either inserting the complete json string into the credentials management system or needing to dynamically construct the json string.  The json string could be constructed on the fly with the external-secrets operator's templating support https://external-secrets.io/latest/guides/templating/ but I'm ignorant about the capabilities of the various other secret injection operators. Would matching `.data` keys heuristically be an option? E.g. `.data.SecretKey_*`?

**Comment by travisn** [comment]
Why does SecretKeys need to be JSON? If the user needs would be updating this field, a comma separated seems easier to work with.

**Comment by BlaineEXE** [comment]
> Why does SecretKeys need to be JSON? If the user needs would be updating this field, a comma separated seems easier to work with.

That is a reasonable question. As long as the delimiter is a non-allowed S3 key character, it should be fine.

Based on this doc, I think commas should work just fine. It seems that `[A-Z][a-z][0-9]/+` (and maybe `;`???) are the characters we could reasonably expect to see based on this doc: https://learn.microsoft.com/en-us/purview/sit-defn-amazon-s3-client-secret-access-key#pattern

**Comment by BlaineEXE** [comment]
Other thoughts for the annotation name:
- secret.rook.io/source-of-truth (obviously would just be for secrets)
- behavior.rook.io/source-of-truth

**Comment by travisn** [comment]
What about this?
```
rook.io/source-of-truth
```
The values would depend on the context where we use it. In this example, the annotation would be set to "secret" to override the default.

**Comment by BlaineEXE** [comment]
Seems reasonable. There are some `kubernetes.io` annotations that don't have a prefix, like these: https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-metadata-name
We don't specifically need a prefix I guess. Let's plan for that then.


I also just had a thought that `rook.io/source-of-truth==secret` effectively means that the user/admin is the owner of the secret. With that in mind, I think it might be good to ensure Rook won't edit/delete the secret when the annotation is present to make sure user changes aren't overwritten. Sound good @travisn ?


With that, I believe this is the finalized plan then:

- Create flow (S3 user doesn't exist yet)
  - if the secret exists, use the k8s secret content as the initial creds for the user
  - if the secret doesn't exist, create a new user with random creds
  - create (or update) the k8s secret with latest credential info from S3
- Update flow (S3 user exists already)
  - if `rook.io/source-of-truth` == `secret`
    - if `SecretKeys` is not empty, verify that `SecretKey` is in the `SecretKeys` list (error if not)
    - update the user creds with the k8s secret content
      - if `SecretKeys` is empty, `SecretKey` is the only key that should exist for the S3 user
      - otherwise, all `SecretKeys` should be active for the S3 user
  - if `rook.io/source-of-truth` != `secret`
    - update the k8s secret with the creds present from the S3 user
      - `SecretKey` should be filled with the last item in list of all active s3 secret keys
      - `SecretKeys` should be filled in with comma-delimited list of all active s3 secret keys
- Delete flow 
  - if `rook.io/source-of-truth` == `secret`, do not delete the secret
  - otherwise, delete the secret

**Comment by parth-gr** [comment]
@BlaineEXE I just don't understand what is ` SecretKeys:` for?
I think it is related to supporting multiple secrets at a time, Let discuss that, maybe we can keep this change straightforward and add this change as a followup for supporting multiple keys

**Comment by BlaineEXE** [comment]
> I just don't understand what is  SecretKeys: for?
I think it is related to supporting multiple secrets at a time, Let discuss that, maybe we can keep this change straightforward and add this change as a followup for supporting multiple keys

You said that when you tried adding a key via `radosgw-admin`, a new key was added, and the old key remained active. If we are beginning to implement better key rotation strategies, it seems natural to me to include some kind of output/feedback that helps users understand that multiple keys may exist. Users may accidentally forget to revoke a compromised key, and this can help them avoid that.

**Comment by parth-gr** [comment]
> - Update flow (S3 user exists already)
>  - if `rook.io/source-of-truth` == `secret`
>    - if `SecretKeys` is not empty, verify that `SecretKey` is in the `SecretKeys` list (error if not)

So the user when updating the secret should update the secret at two places always the  `SecretKey` and then `SecretKeys` list ?  

>    - update the user creds with the k8s secret content
      - if `SecretKeys` is empty, `SecretKey` is the only key that should exist for the S3 user
      - otherwise, all `SecretKeys` should be active for the S3 user

>  - if `rook.io/source-of-truth` != `secret`
>    - update the k8s secret with the creds present from the S3 user
 >     - `SecretKey` should be filled with the last item in list of all active s3 secret keys
 
 Do you mean the latest key, actually it appends the latest key at the top of the array

>       - `SecretKeys` should be filled in with comma-delimited list of all active s3 secret keys

 
      if the `SecretKeys` already has some values that should be deleted and replaced with the ceph user keys?

**Comment by BlaineEXE** [comment]
I've been investigating key rotation for the COSI design updates as well, and after speaking with Josh on a call and collecting more information, I think we might want to revisit the design discussion for this Rook feature as well. It's tiring to get 90% of the way to completion with design and code and then have to rework it, but I think it may be worthwhile in this case.

From my discussion with Josh, the biggest gap that I think we have with the implementation proposed is that users don't have a good way to determine when the key rotation has completed. Josh has over a dozen k8s clusters that he manages, and that may grow to 2 dozen in not too long. In large environments like that, being able to know definitively when a key has been rotated is important. Status/monitoring/feedback is also a major tenet of the k8s design philosophy as well, and I missed it when proposing a design for this feature.

I'd like to talk about the details more in a Rook huddle next week, but the summary of my proposed new design is to not reuse the current secret for both input and output. I think it's simpler for developers and for users if we use a dedicated input secret, and Rook always uses the existing secret for output.

As for monitoring, I propose to achieve that by recording the `metadata.generation` field of the user input secret as a status field on the COSUser CR. Users can then easily use k8s tooling to compare the two values to know when the key rotation is complete. 

I'd like to discuss this as a general idea before we start changing the code, so we don't burn ourselves out when reworking this. I think we can reuse a large part of the work we've already done on specifying the secret formatting, so hopefully we can limit that aspect. If that sounds good, we can iron out smaller details regarding COSUser spec/status API fields.

**Comment by parth-gr** [comment]
So for implementation:

1) Keep the original rgw secret as it is for the status.

2) If a user wants to provide their own credentials or do key rotation.
       > Provide the user secret name in the objectuser CR
          CR will have a spec,
          ```
          spec:
             userSecretName: my-secret
         ``` 
       > Create or update the secret with this name
       ```
            apiVersion: v1
            kind: Secret
            metadata:
                 name: my-secret
            data:
              AccessKey: U0lVNVJQTVg0VFIwMTI1QkxGSjc=
              SecretKey: aHdnbGdIVmlDcEJqTHN3WUJDakJqWlBRQ2tzeEM3dkRuRDdXMTNySw==
              Endpoint: aHR0cDovL3Jvb2stY2VwaC1yZ3ctbXktc3RvcmUucm9vay1jZXBoLnN2Yzo4MA==
       ```
       > The data fields remain the same as it was before
       > If the reconciliation of cephobjectstore user was successful and credentials were updated, 
             > Add the status to the cephobjectstore user CR, `my-user secret creds successfully updated`
             > Also update the `original rgw  secret` with the `AccessKey` and `SecretKey`.
             
  @BlaineEXE please review is that what we need to do?

**Comment by BlaineEXE** [comment]
The key reason to make this change is to make sure we are also planning a way for users to get feedback about the key rotation process. We need to also plan the `status` fields. I would suggest this to be something like this.

```
status:
  # ...
  inputCredentials:
    secretName: <the name of the secret that Rook last used to apply creds>
    secretGeneration: <the generation of the secret Rook last used to apply creds>
```

Having secretName and secretGeneration are both important. The first allows the user to change the secret used, and ensure Rook applied the correct one. The second allows the user to know when the current secret matches what Rook has applied.

From my discussion with Josh, I think it is also still good to plan for a way that users can have multiple active credentials. Josh explained this to be important to him, as well as other admins. It allows a migration window for end users so that key rotation isn't something that causes application downtime.

I think we should still plan to include the `Credentials` field in both the output secret and input secret to allow for this. (We could use multiple input credentials, but I think having a single input secret will mean that status comparison is easier for users.)

Also, I have another concern: Because the resource is called CephObjectStore**User**, I think we should refrain from choosing **user**SecretName. IMO, this naming overlap allows too much room for misunderstanding.

I also think we might consider using the a new section and the `inputCredentials.secretName` pattern for the input spec. This helps clarify for users that we are inputting credentials, and that it takes a secret name. I don't think this is strictly necessary, but by creating a new section, it allows us flexibility in the future to add more fields to `inputCredentials` if we need to, without bloating the main `spec`. I'm curious to hear @travisn 's preferences here. I could go either way.

If we keep it in the top level, `inputCredentialsSecretName` would be the most clear naming, though `inputSecretName` would probably be fine. I would steer clear of `credentialsSecretName` since that might also be interpreted as the config changes the name of the secret Rook uses for output.

**Comment by travisn** [comment]
If we're going to use a different secret for input, that needs to be on the spec, right? Then the status would just have the generation of that secret that has been evaluated, something like this?
```
spec:
  inputCredentials:
    secretName: <the name of the secret that Rook last used to apply creds>
status:
  inputCredentials:
    secretGeneration: <the generation of the secret Rook last used to apply creds>
```

And I'm not clear what the proposal is for specifying multiple sets of creds? For simplicity it's nice to have the secret values match the output secret values, but if multiple sets of creds need to be specified, how are additional creds specified?

**Comment by parth-gr** [comment]
@BlaineEXE user secret should not provide `Endpoint` data, right?

**Comment by BlaineEXE** [comment]
> BlaineEXE user secret should not provide `Endpoint` data, right?

Correct. The input secret should only contain credentials.



> If we're going to use a different secret for input, that needs to be on the spec, right? Then the status would just have the generation of that secret that has been evaluated, something like this?
> 
> ```
> spec:
>   inputCredentials:
>     secretName: <the name of the secret that Rook last used to apply creds>
> status:
>   inputCredentials:
>     secretGeneration: <the generation of the secret Rook last used to apply creds>
> ```
> 
> And I'm not clear what the proposal is for specifying multiple sets of creds? For simplicity it's nice to have the secret values match the output secret values, but if multiple sets of creds need to be specified, how are additional creds specified?

The secret name should also be reflected in the status in case the user changes the secret name in the spec. If a user creates a new secret each time, then the generation may always be 1, and there will be ambiguity as to whether the correct credentials are applied.

```
spec:
  inputCredentials:
    secretName: <the name of the secret the user wishes credentials to be applied from>
status:
  inputCredentials:
    secretName: <the name of the secret that Rook last used to apply creds>
    secretGeneration: <the generation of the secret Rook last used to apply creds>
```

**Comment by parth-gr** [comment]
@BlaineEXE some more questions on the status field,

```
status:
  inputCredentials:
    secretName: <the name of the secret that Rook last used to apply creds>
    secretGeneration: <the generation of the secret Rook last used to apply creds>
```

So wanted to understand what should be the status on different events,

1) If the reconciliation is started and secret exists -> secretName: my-secret, secretGeneration:-1
2) If the reconciliation is completed succesfully and secret exists -> secretName: my-secret, secretGeneration: (generation value of secret)
3) If the reconciliation is completed with a failure and secret exists -> secretName: my-secret, secretGeneration: (not sure?)
4) If the reconciliation is completed succesfully and the secret doesn't exists -> secretName: -1, secretGeneration:-1
5) If the reconciliation is started and the secret doesn't exists -> secretName: -1, secretGeneration:-1
6) If the reconciliation is completed with failure and the secret doesn't exists -> secretName: -1, secretGeneration:-1

**Comment by BlaineEXE** [comment]
These are great questions. :)

The status should reflect what the current reconcile has completed, unless the status item is specifically intended to reflect something about the ongoing reconcile. Since we aren't telling the user about the reconcile, it means we should update this status only after modifying things related to it.

When planning for the behavior of when this particular data gets updated in the status, we need to consider all cases:
1. user changes `inputCredentials.secretName` to a different value
    - special case is when the user first sets this: i.e., "" -> "something"
    - different special case is when the user removes this: i.e., "something" -> ""
2. user modifies the secret itself
3. nothing secret-related has changed (i.e. a CephObjectStoreUser change that is unrelated to the secret)

Reconciliation routines work best when the logic used in code isn't unique for any of the cases. In this case, there is no need for us to use special logic for any of the cases. But we do need to consider all 3 of the cases I mentioned when deciding how to implement the behavior.

> * If the reconciliation is started and secret exists -> secretName: my-secret, secretGeneration:-1

We should not update these status items like this on every reconcile. In the no.3 case where nothing secret-related has changed, changing the status like this will only confuse users.

IMO, we can consider changing the status to be this immediately before Rook attempts to update the credentials, for cases no.1 and no.2. Doing so would be an indicator of what Rook is attempting to do (update the creds), and communicates that the underlying user credential status is undefined. However, once we update these statuses to reflect undefined internal state, we can't go back.

In my experience, sometimes the RGW will time out or will return an error but will still have made some internal changes. I think it would be good for Rook to indicate undefined state like this, but we should only set it immediately before RGW credentials are going to be modified.

> * If the reconciliation is completed succesfully and secret exists -> secretName: my-secret, secretGeneration: (generation value of secret)

Yes. But let's also be clear that this also assumes the user has specified an input secret. And be sure to remember that this should not be the values of the secret at the time the status is applied. The fields should reflect the secret name and generation that were applied to Ceph, even if they are behind the current secret name/value.

> * If the reconciliation is completed with a failure and secret exists -> secretName: my-secret, secretGeneration: (not sure?)

Let's also be clear that this also assumes the user has specified an input secret.

Failures can have different sources. If the reconcile fails before Rook attempts to update the credentials, there is no need to change these status fields.

If there is an error while applying the user secret data to the user, then we should not update the fields either. This is because we have already (presumably) set the status such that it indicates an undefined result. When RGW says that it failed to do the update, we don't truly know what the internal state is. 

> * If the reconciliation is completed succesfully and the secret doesn't exists -> secretName: -1, secretGeneration:-1

Mostly yes, but I would use more straightforward wording: if the reconciliation is completed, and **the user hasn't specified `inputCredentials.secretName`**. Also, in the case where the user hasn't set (or has recently removed the inputs), the status should just be the default value for both items, so: secretName: "", secretGeneration: 0

> * If the reconciliation is started and the secret doesn't exists -> secretName: -1, secretGeneration:-1

And the user has specified an input secret.

This is an interesting case. When thinking about this, I am trying to determine what will be the best for the user experience. Presumably, the user will have an alert that checks whether the status Rook is outputting matches the secret used for input creds.

With that case in mind, it probably is good to set these such that they will throw an error Rook can't clear up the issue.

However, case no.1 is the complicating one. What happens if the user changes the value from "secret-1" to "secret-2"?

From a behavioral standpoint, Rook isn't going to be able to apply creds from a non-existent secret. That means that the underlying user's credentials are known by Rook, and they are known to be the currently-set values. Rook should not update the status in this case. The status should remain at secretName: "secret-1", secretGeneration: \<previous-value\>

> * If the reconciliation is completed with failure and the secret doesn't exists -> secretName: -1, secretGeneration:-1

This shouldn't happen, and we don't need to handle this condition. If the secret doesn't exist, that is already a failure.

-----

I am trying to focus on and push all of the Rook devs to write less complex and fully-idempotent reconciliation routines and reconciliation logic. With that in mind, I will propose some pseudocode below that I think matches the enumerated cases above.

These are good focus points:
- use as few conditional branches as possible
- conditional branches should be as short as possible (keep happy path on the left)
- conditionals should largely be based on (1) input config, and (2) error status
- avoid conditionals based on edge conditions where possible
  - this specially goes for the edge condition "if this is the first reconcile"; it should be possible to encode logic for the first, Nth, and last times a feature is used using the same code logic
- reconciliation routines should always refresh their current "view" of the inputs and system states every reconcile -- reconciles should not store state/status that gets used in future reconciles


```
ReconcileCredentials()
  if spec.inputCredentials.inputSecret is set:
     if inputSecret doesn't exist:
       return error // we know that the internal state is still the previously set secret

     secretObj, error = getSecret(inputSecret)
     if error:
       return error // we know that the internal state is still the previously set secret
        
     // once rook begins to update the user creds, we assume an undefined internal state
     // even if RGW returns failure (or timeout), sometimes there is a partial internal update
     set status.secretName = inputSecret, status.secretGeneration=-1 
     
     error = setUserCreds(user, creds(secretObj.data))
     if error:
        return error // no need to change status since undefined state will indicate a problem
  endif
  
  // regardless of whether the input secret is used, Rook always gets the current internal creds to apply to the reference secret
  currentCreds = getUserCreds(user)
  error = updateReferenceSecret(currentCreds)
  if error:
    // in case where inputSecret is present, there is no need to change status since undefined state will indicate a problem
    return error 
  
  if spec.inputCredentials.inputSecret is not set:
    set status.secretName = "", staus.secretGeneration=0
    return success
    
  if currentCreds != creds(inutSecret.data):
    return error(input secret was not applied successfully) // keep undefined status to indicate a problem
    
  set status.secretName = inputSecret, status.secretGeneration = secretObj.generation
  
  return success
```

Notice that we don't actually have to have conditionals for every one of the cases you have mentioned. The code itself is simpler. But it is still a good exercise to enumerate all of the different input and error conditions beforehand, so we know whether the code we produce is doing the right thing in all cases. It is also good practice to indicate in code comments what Rook is choosing to do (or not to do) based on the enumerated conditions we planned.

**Comment by BlaineEXE** [comment]
I did some looking locally, and it looks like not all k8s resources utilize the `metadata.generation` field. Secrets in particular seem to not update this. We will have to use another metric for determining if Rook has applied the current version. 

The next best candidate is `resourceVersion`. This is [documented to be](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes)
> the version of that resource as stored in the underlying persistence layer

That doesn't make any special guarantees that the field won't change when the user hasn't made any updates, but it seems like the persistently stored version shouldn't change randomly, especially of a Secret, which doesn't have a status field. For resources that have a Status, I would bet that a Status change changes the `resourceVersion` but doesn't change the `generation`.

In my test cluster, I haven't seen a secret randomly get a new `resourceVersion` after 30+ minutes.

**Comment by cfis** [comment]
It would be great if this could be finished off - we very much would like to have this feature. In our code base, s3 credentials are used in many places (rook ceph of course, argo workflows for artifacts, custom services to pull files, algorithms to pull files, user uploads to save files, etc). It would be much more convenient to have a s3 credentials secret and replicate it between namespaces then have to special case ceph (since first you have to install ceph, then get the generated access/secret keys, then update a secret). Happy to test when its ready.

**Comment by jhoblitt** [comment]
Now that it is a different year and a different dev looking at the implementation... I'm thinking that a good resolution to key rotation is to explicitly list all keys set on the user.  IOW - any key not specified in the list would be purged.  One of the main advantages to this approach is no annotations/labels would be required on the secret resources and it is relatively straight forward to implement idempotently as the full list of keys would be explicit on the `CephObjectStoreUser`. As the secrets containing the keys are enumerated, the data keys in the secret could be explicitly set as well and the secret resource would not have to be in a specific format.  E.g.:
```yaml
---
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: foo
  namespace: rook-ceph
spec:
  store: lfa
  clusterNamespace: rook-ceph
  quotas:
    maxBuckets: 1
  keys:
    - acccessKey:
        secretKeyRef:
          name: foo-s3
          key: AWS_ACCESS_KEY_ID
      secretKey:
        secretKeyRef:
          name: foo-s3
          key: AWS_SECRET_ACCESS_KEY
    - acccessKey:
        secretKeyRef:
          name: bar-s3
          key: AWS_ACCESS_KEY_ID
      secretKey:
        secretKeyRef:
          name: bar-s3
          key: AWS_SECRET_ACCESS_KEY
```

**Comment by jhoblitt** [comment]
Do we need to watch secrets to trigger reconciliation or is it sufficient to only reconcile when the `CephObjectStoreUser` is updated (or the operator is restarted)?

**Comment by parth-gr** [comment]
@jhoblitt are you picking this item?
I would like to re-work on the approach from where we started https://github.com/rook/rook/pull/14001, I do have certain free cycles this quarter.

**Comment by BlaineEXE** [comment]
```yaml
  keys:
    - acccessKey:
        secretKeyRef:
          name: foo-s3
          key: AWS_ACCESS_KEY_ID
      secretKey:
        secretKeyRef:
          name: foo-s3
          key: AWS_SECRET_ACCESS_KEY
    - acccessKey:
        secretKeyRef:
          name: bar-s3
          key: AWS_ACCESS_KEY_ID
      secretKey:
        secretKeyRef:
          name: bar-s3
          key: AWS_SECRET_ACCESS_KEY
```

This API would change the things we talked about and agreed upon a while ago. I'm not opposed to it, but we'll need to discuss that it looks good to everyone. The biggest change this will cause in cascade is needing to track each secret's `resourceVersion` separately in the status like this:

```yaml
inputCredentials:
  secrets:
    - name: foo-s3
      resourceVersion: N
    - name: bar-s3
      resourceVersion: M
    # this could get a bit unweildy if access/secret key all use different secrets
```

Personally, it seems like it would be challenging as a user to write an alert that warns me if the input secret and output secret aren't matching. Maybe I'm wrong about that because I rarely have to write things like that, but I do want to make sure the monitoring and alerting experience is workable here.

A way to meet in the middle would be to take the assumption that a paired secretkey and accesskey are always in the same secret. IMO, that is a reasonable assumption, and would help simplify things a bit and help align the input and status data.

Something like this:
```yaml
inputCredentials: # this is wording I'd like to keep, following the long discussion above
  secrets:
    - name: foo-s3
      secretKeyDataField: AWS_ACCESS_KEY_ID # assume as default val to make this optional?
      accessKeyDataField: AWS_SECRET_ACCESS_KEY # assume as default val to make this optional?
```

> I'm thinking that a good resolution to key rotation is to explicitly list all keys set on the user. IOW - any key not specified in the list would be purged.

I totally agree. Having extraneous keys not specified by the input secret (source of truth) doesn't make sense to me. 

> Do we need to watch secrets to trigger reconciliation or is it sufficient to only reconcile when the CephObjectStoreUser is updated (or the operator is restarted)?

IMO, the complete solution is to watch secrets, but I don't think that has to be in the MVP. That can be a future item.

**Comment by jhoblitt** [comment]
> Personally, it seems like it would be challenging as a user to write an alert that warns me if the input secret and output secret aren't matching. Maybe I'm wrong about that because I rarely have to write things like that, but I do want to make sure the monitoring and alerting experience is workable here.

I have not tried to write such a rule.  It appears that `kube-state-metrics` exports `kube_secret_metadata_resource_version` out of the box, so it is likely possible to write a `prometheusrule` if there was an exporter that exposed the details of `cephobjectstoreuser`.  My thinking on how to handle checking multiple sets of credentials being active at the same time would be to add s3 auth support to blackbox exporter or to write a custom exporter with that functionality (blackbox already supports http basic and oauth so v4 signatures might be accepted)... but I do not currently have a solution.

 > A way to meet in the middle would be to take the assumption that a paired secretkey and accesskey are always in the same secret. IMO, that is a reasonable assumption, and would help simplify things a bit and help align the input and status data.

My thinking was that `secretKeyRef` is a fairly common field name that most k8s users understand the semantics of.  E.g. `env[].valueFrom.secretKeyRef` in pods but I'm agreeable to different field names.
> 
> Something like this:
> 
> inputCredentials: # this is wording I'd like to keep, following the long discussion above
I proposed `keys` as it is consistent with the output of `radosgw-admin`. E.g.
```
bash-5.1$ radosgw-admin user info --uid test
{
    "user_id": "test",
    "display_name": "test1",
    "email": "",
    "suspended": 0,
    "max_buckets": 1000,
    "subusers": [],
    "keys": [
        {
            "user": "test",
            "access_key": "XXX",
            "secret_key": "XXX"
        }
    ],
    "swift_keys": [],
...
```
and the `radosgw-admin key` sub-command. This would probably be intuitive naming to most ceph admins.
(it also aligns with the names of funcs in go-ceph for [managing credentials](https://github.com/ceph/go-ceph/blob/master/rgw/admin/keys.go#L12))

>   secrets:
>     - name: foo-s3
>       secretKeyDataField: AWS_ACCESS_KEY_ID # assume as default val to make this optional?
>       accessKeyDataField: AWS_SECRET_ACCESS_KEY # assume as default val to make this optional?

I would rather start off with requiring explicit field names as that is one less behavior that compatibility has to be maintained for and thus one less acceptance test to write / run.

> IMO, the complete solution is to watch secrets, but I don't think that has to be in the MVP. That can be a future item.
👍

**Comment by jhoblitt** [comment]
I wasn't aware of it but it looks like CRs can be exposed via `kube-state-metrics`: https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md

**Comment by jhoblitt** [comment]
My initial proposal could be simplified to: 
```yaml
---
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: foo
  namespace: rook-ceph
spec:
  store: lfa
  clusterNamespace: rook-ceph
  quotas:
    maxBuckets: 1
  keys:
    - acccessKeyRef
        name: foo-s3
        key: AWS_ACCESS_KEY_ID
      secretKeyRef
        name: foo-s3
        key: AWS_SECRET_ACCESS_KEY
    - acccessKeyRef
        name: bar-s3
        key: AWS_ACCESS_KEY_ID
      secretKeyRef
        name: bar-s3
        key: AWS_SECRET_ACCESS_KEY
```

Which can be defined satisfyingly succinctly as:

```go
diff --git a/pkg/apis/ceph.rook.io/v1/types.go b/pkg/apis/ceph.rook.io/v1/types.go
index de8c70cce..46bda43f9 100755
--- a/pkg/apis/ceph.rook.io/v1/types.go
+++ b/pkg/apis/ceph.rook.io/v1/types.go
@@ -19,6 +19,7 @@ package v1
 import (
        "time"
 
+       corev1 "k8s.io/api/core/v1"
        v1 "k8s.io/api/core/v1"
        "k8s.io/apimachinery/pkg/api/resource"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -1996,6 +1997,10 @@ type ObjectStoreUserSpec struct {
        // +optional
        // +nullable
        Quotas *ObjectUserQuotaSpec `json:"quotas,omitempty"`
+       // Allows specifying credentials for the user. If not provided, the operator
+       // will generate them.
+       // +optional
+       Keys []ObjectUserKey `json:"keys,omitempty"`
        // The namespace where the parent CephCluster and CephObjectStore are found
        // +optional
        ClusterNamespace string `json:"clusterNamespace,omitempty"`
@@ -2086,6 +2091,15 @@ type ObjectUserQuotaSpec struct {
        MaxObjects *int64 `json:"maxObjects,omitempty"`
 }
 
+// ObjectUserKey defines a set of rgw user access credentials to be retrived
+// from secret resources.
+type ObjectUserKey struct {
+       // Secret key selector for the access_key (commonly referred to as AWS_ACCESS_KEY_ID).
+       AccessKeyRef *corev1.SecretKeySelector `json:"AccessKeyRef,omitempty"`
+       // Secret key selector for the secret_key (commonly referred to as AWS_SECRET_ACCESS_KEY).
+       SecretKeyRef *corev1.SecretKeySelector `json:"SecretKeyRef,omitempty"`
+}
+
 // +genclient
 // +genclient:noStatus
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```

**Comment by cfis** [comment]
So how would this work? Rook no longer creates a default secret? Instead users provide secret refs, which then need to get tieds to Object Store uses on their creation (or can be added later)?

**Comment by jhoblitt** [comment]
> So how would this work? Rook no longer creates a default secret? Instead users provide secret refs, which then need to get tieds to Object Store uses on their creation (or can be added later)?

The current behavior would be retained when `.spec.keys` (name TBD) is empty. Backwards compatibility is @travisn's prime directive.

**Comment by jhoblitt** [comment]
Here is a stand alone proof of concept of using go-ceph to reconcile multiple s3 keys at once: https://gist.github.com/jhoblitt/f87bb406344bdd2ce1770fb14299d8d7

**Comment by BlaineEXE** [comment]
> > Personally, it seems like it would be challenging as a user to write an alert that warns me if the input secret and output secret aren't matching. Maybe I'm wrong about that because I rarely have to write things like that, but I do want to make sure the monitoring and alerting experience is workable here.
> 
> I have not tried to write such a rule. It appears that `kube-state-metrics` exports `kube_secret_metadata_resource_version` out of the box, so it is likely possible to write a `prometheusrule` if there was an exporter that exposed the details of `cephobjectstoreuser`. My thinking on how to handle checking multiple sets of credentials being active at the same time would be to add s3 auth support to blackbox exporter or to write a custom exporter with that functionality (blackbox already supports http basic and oauth so v4 signatures might be accepted)... but I do not currently have a solution.

We need a way to track that the input secrets are getting applied in a human-readable way. Relying on prometheus metrics for detecting that the system is operating correctly seems like it's missing something critical. I want to make sure I and users are able to look at the input and secrets and the status and see that things are updated or not.

**Comment by cfis** [comment]
Thanks for working on this. Any plans on getting this into a future release?

**Comment by richardelling** [comment]
I'm a little bit confused, in part due to the overloading of the term "secret": kubernetes secrets and S3 secret keys. Is my understanding correct that
```
  keys:
    - acccessKeyRef
        name: foo-s3
        key: AWS_ACCESS_KEY_ID
      secretKeyRef
        name: foo-s3
        key: AWS_SECRET_ACCESS_KEY
```
means that `keys[0].secretKeyRef.name` is a name for a kubernetes Secret? If so, wouldn't it be clearer to use the key `secretName` rather than `name` (as used elsewhere where resources reference secrets)?

**Comment by jhoblitt** [comment]
> Thanks for working on this. Any plans on getting this into a future release?

There's currently an open PR that is working but review comments need to be addressed: https://github.com/rook/rook/pull/15359

**Comment by jhoblitt** [comment]
> I'm a little bit confused, in part due to the overloading of the term "secret": kubernetes secrets and S3 secret keys. Is my understanding correct that
> 
> ```
>   keys:
>     - acccessKeyRef
>         name: foo-s3
>         key: AWS_ACCESS_KEY_ID
>       secretKeyRef
>         name: foo-s3
>         key: AWS_SECRET_ACCESS_KEY
> ```
> 
> means that `keys[0].secretKeyRef.name` is a name for a kubernetes Secret? If so, wouldn't it be clearer to use the key `secretName` rather than `name` (as used elsewhere where resources reference secrets)?

If you have ever defined an env var in a pod from a secret, the semantics would be the same. In fact, the implimentation in #15359 is the using the exact same type.  E.g.:

```console
 ~ $ k explain pod.spec.containers.env.valueFrom.secretKeyRef
KIND:       Pod
VERSION:    v1

FIELD: secretKeyRef <SecretKeySelector>

DESCRIPTION:
    Selects a key of a secret in the pod's namespace
    SecretKeySelector selects a key of a Secret.
    
FIELDS:
  key	<string> -required-
    The key of the secret to select from.  Must be a valid secret key.

  name	<string>
    Name of the referent. This field is effectively required, but due to
    backwards compatibility is allowed to be empty. Instances of this type with
    an empty value here are almost certainly wrong. More info:
    https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names

  optional	<boolean>
    Specify whether the Secret or its key must be defined
```

**Comment by jhoblitt** [comment]
> > > Personally, it seems like it would be challenging as a user to write an alert that warns me if the input secret and output secret aren't matching. Maybe I'm wrong about that because I rarely have to write things like that, but I do want to make sure the monitoring and alerting experience is workable here.
> > 
> > 
> > I have not tried to write such a rule. It appears that `kube-state-metrics` exports `kube_secret_metadata_resource_version` out of the box, so it is likely possible to write a `prometheusrule` if there was an exporter that exposed the details of `cephobjectstoreuser`. My thinking on how to handle checking multiple sets of credentials being active at the same time would be to add s3 auth support to blackbox exporter or to write a custom exporter with that functionality (blackbox already supports http basic and oauth so v4 signatures might be accepted)... but I do not currently have a solution.
> 
> We need a way to track that the input secrets are getting applied in a human-readable way. Relying on prometheus metrics for detecting that the system is operating correctly seems like it's missing something critical. I want to make sure I and users are able to look at the input and secrets and the status and see that things are updated or not.

The possible method of generating alerts with Prometheus that I outlined was specifically in response to your concern that it needs to be possible to do alerting (which I believe it is).

Currently, when a `CephObjectStoreUser` reconciliation fails `.status.phase` is set to `ReconcileFailed`.  That behavior isn't changed in the current iteration of https://github.com/rook/rook/pull/15359. A failure to reconcile keys triggers the same error handling chain that results in a the phase be set to `ReconcileFailed`.  Are you suggesting that a different behavior is needed?

**Comment by jhoblitt** [comment]
> Currently, when a `CephObjectStoreUser` reconciliation fails `.status.phase` is set to `ReconcileFailed`. That behavior isn't changed in the current iteration of [#15359](https://github.com/rook/rook/pull/15359). A failure to reconcile keys triggers the same error handling chain that results in a the phase be set to `ReconcileFailed`

I decided to write an integration test to ensure ReconcileFailed is being set and it appears there are many ways unrelated to #15359 in which the reconcile can fail without updating `.status.phase`.  Ugh.

**Comment by jhoblitt** [comment]
> I decided to write an integration test to ensure ReconcileFailed is being set and it appears there are many ways unrelated to [#15359](https://github.com/rook/rook/pull/15359) in which the reconcile can fail without updating `.status.phase`. Ugh.

Possible fix for updating the CR status when reconcile fails: #15530.

**Comment by cfis** [comment]
Thanks for doing this!
